docs/configuration.md
# Configuration

SPT offers several methods of configuration, ranging from required, modifying existing functionality, and fully adding brand new behavior.

## Variables

The set of variables that are configurable within the system can be set in a combination of two ways:
  * The configuration File (`app.yaml`)
  * Environment Variables

Either way these variables are set they are only ever read on initial startup of the application, meaning if they've changed while the system is running, it'll need to be restarted for the changes to take effect.
The full list of values can be viewed in the example [`app.yaml`](../app.example.yaml) along with most of their default values.

Either can be used to modify the following values:

* `PORT`: This is the port the SPT will be exposed on. (If running in Docker, this is the Docker Containers Exposed Port). **Default `8080`**
* `SERVER_URL`: The URL the server will assume it is located at. Used when creating callbacks via OAuth, and when supplying pagination information. **Default `"http://localhost:8080"`**
* `DB_HOST`: This is the host of the PostgreSQL database. **No Default.**
* `DB_USER`: This is the username to login to the PostgreSQL database. **No Default.**
* `DB_PASS`: This is the password to login to the PostgreSQL database. **No Default.**
* `DB_DB`: This is the database to connect to within the PostgreSQL server. **No Default.**
* `GOOGLE_CLIENT_ID`: This is the Google App Client ID for OAuth. **No Default.**
* `GOOGLE_CLIENT_SECRET`: This is the Google App Client Secret for OAuth. **No Default.**
* `RATE_LIMIT_GENERIC`: This is the generic rate limit applied. Should be a value of requests per IP per 15 minutes window. **Default 1200.**
* `PAGINATION_LIMIT`: The value long returns are paginated by. **Default 30**
* `SITE_NAME`: A human friendly name for the website. **Default "Student Point Tracker"**
* `CSV_DELIMITER`: The delimiter used within CSV's being used for importing users. **Default ","**
* `DOMAIN`: The email domain being used eg. "example.edu". **Default "The email domain used."**
* `CACHE_TIME`: The time in milliseconds to cache information server-side. **Default 604800**
* `SESSION_FILE_STORE_TTL`: The time-to-live on the session file store. Essentially how long a user can stay logged in. **Default 3600.**
* `REQUIRE_LOGIN`: A boolean which is controls if a user **must** be signed in at all times. **Default true**
* `REPORT_A_PROBLEM_URL`: The URL to link to on the footer of every page for users to report a problem. **Default "https://github.com/confused-Techie/student-point-tracker/issues"**
* `LOCALE`: Your user base's locale. **Default "en-US"**
* `REDIRECT_STUDENTS`: If students should be redirected from the home page to their account page. **Default true**
* `FOTER_ITEM_NAME`: The custom footer element's name. **Default false**.
* `FOOTER_ITEM_LINK`: The custom footer element link. **Default false**
* `STARTUP_DB_CONNECT_RETRY_COUNT`: The amount of attempts that will be made to connect to the database during application startup. **Default 10**
* `STARTUP_DB_CONNECT_RETRY_TIME_MS`: The time in milliseconds to wait between each attempt to connect to the database during startup. **Default 1000**
* `DEV_LOGIN`: **For Development** The developer login email address.
* `DEV_IS_STUDENT`: **For Development** If the developer login should be interpreted as a student account.
* `DEV_IS_ADMIN`: **For Development** If the developer login should be interpreted as an admin account.
* `ADMINS`: An array of admin email addresses for the system.
* `POINT_CHIPS`: An array of numerical point values, to define the different values of the "Quick Points". **No Default**
* `TASKS`: An array of task objects, to define the different "Tasks". **No Default**
* `COLUMNS`: A value for how to interprete the columns in the import CSV, as defined [here](https://csv.js.org/parse/options/columns/). **Default true**

## Environment Variables

While the above variables can be optionally configured via environment variables or via the `app.yaml` configuration file, the following values are ones that can **only** be configured via environment variables.

* `STP_RESOURCE_PATH`: This is the path on the filesystem that points to the location of STP's root directory for storing information, finding configuration files etc. **Default "./storage"**


src/config.js
const fs = require("fs");
const path = require("path");
const yaml = require("js-yaml");

function findResourcePath() {
  return process.resourcePath ?? process.env.STP_RESOURCE_PATH ?? "./storage";
}

function getConfigFile() {
  try {
    let data = null;

    let configLoc = findResourcePath();
    let fileContent = fs.readFileSync(path.join(configLoc, "app.yaml"), "utf8");
    data = yaml.load(fileContent);

    return data;
  } catch (err) {
    if (process.env.PROD_STATUS === "dev") {
      console.error(
        `Failed to load app.yaml in non-production env! ${err.toString()}`
      );
      data = {};

      return data;
    } else {
      console.error("Failed to load app.yaml!");
      console.error(err);
      process.exit(1);
    }
  }
}

function getConfig() {
  let data = getConfigFile();

  const findValue = (key, def) => {
    return process.env[key] ?? data[key] ?? def;
  };

  // Now our config is a JavaScript Object
  // And we will now use a custom object here to return all values
  return {
    PORT: findValue("PORT", 8080),
    SERVER_URL: findValue("SERVER_URL", "http://localhost:8080"),
    DB_HOST: findValue("DB_HOST"),
    DB_USER: findValue("DB_USER"),
    DB_PASS: findValue("DB_PASS"),
    DB_DB: findValue("DB_DB"),
    DB_PORT: findValue("DB_PORT"),
    GOOGLE_CLIENT_ID: findValue("GOOGLE_CLIENT_ID"),
    GOOGLE_CLIENT_SECRET: findValue("GOOGLE_CLIENT_SECRET"),
    RATE_LIMIT_GENERIC: findValue("RATE_LIMIT_GENERIC", 1200),
    PAGINATION_LIMIT: findValue("PAGINATION_LIMIT", 30),
    SITE_NAME: findValue("SITE_NAME"),
    CSV_DELIMITER: findValue("CSV_DELIMITER", ","),
    DOMAIN: findValue("DOMAIN"),
    CACHE_TIME: findValue("CACHE_TIME", 604800),
    POINT_CHIPS: findValue("POINT_CHIPS", []),
    ADMINS: findValue("ADMINS", []),
    TASKS: findValue("TASKS", []),
    SESSION_FILE_STORE_TTL: findValue("SESSION_FILE_STORE_TTL", 3600),
    DEV_LOGIN: findValue("DEV_LOGIN", null),
    DEV_IS_ADMIN: findValue("DEV_IS_ADMIN", false),
    DEV_IS_STUDENT: findValue("DEV_IS_STUDENT", false),
    REQUIRE_LOGIN: findValue("REQUIRE_LOGIN", true),
    REPORT_A_PROBLEM_URL: findValue(
      "REPORT_A_PROBLEM_URL",
      "https://github.com/confused-Techie/student-point-tracker/issues"
    ),
    LOCALE: findValue("LOCALE", "en-US"),
    REDIRECT_STUDENTS: findValue("REDIRECT_STUDENTS", false),
    FOOTER_ITEM_NAME: findValue("FOOTER_ITEM_NAME", false),
    FOOTER_ITEM_LINK: findValue("FOOTER_ITEM_LINK", false),
    STARTUP_DB_CONNECT_RETRY_COUNT: findValue(
      "STARTUP_DB_CONNECT_RETRY_COUNT",
      10
    ),
    STARTUP_DB_CONNECT_RETRY_TIME_MS: findValue(
      "STARTUP_DB_CONNECT_RETRY_TIME_MS",
      1000
    ),
    RESOURCE_PATH: findResourcePath(),
    COLUMNS: findValue("COLUMNS", true),
  };
}

module.exports = getConfig;
